### ISSUE TITLE
Performance of useSyncExternalStore and React transitions

### ISSUE THREAD

**Issue by OliverJAsh** [reduxjs/react-redux#2086]
### What version of React, ReactDOM/React Native, Redux, and React Redux are you using?

- React: 18.2.0 (latest at time of writing)
- Redux: 4.2.1 (latest at time of writing)
- React Redux: 8.1.3 (latest at time of writing)

### What is the current behavior?

Reduced test case: https://github.com/OliverJAsh/react-redux-useSyncExternalStore-perf

Despite wrapping calls to `dispatch` with `startTransition`, React will always render Redux state updates synchronously. This is in contrast to regular React state updates.

I suspect this is due to the behaviour of `useSyncExternalStore`. However I was wondering if you're aware of any plans for the React team to address this?

As it stands I think this is a significant performance issue. For example, here's a performance recording from Unsplash (the application I work on). We're able to benefit from concurrent rendering and "time slicing" for hydration, but then once hydration has finished we need to run some state updates through Redux, and this results in this very long task (hovered):

![image](https://github.com/reduxjs/react-redux/assets/921609/6e0b0f97-2ea1-4c25-b369-5f0e47e153a8)

It would be great if we could benefit from transitions / concurrent rendering / time slicing.

In our case at Unsplash, if we could figure out a way to make these states updates non-blocking then we could significantly improve our page load performance and metrics such as TBT, TTI, and ITP.

If it helps, here's a demo of my reduced test case, comparing state updates for React vs Redux (both wrapped in `startTransition`):

https://github.com/reduxjs/react-redux/assets/921609/27d767e3-dfb1-469c-9c2c-f64a8b1e9038

### What is the expected behavior?

N/A

### Did this work in previous versions of React Redux?

- [ ] Yes

**Comment by phryneas** [comment]
It is pretty much the React team's decision to deal with all external data sources like this - they created `useSyncExternalStore` for exactly the purpose of keeping "non-React" updates synchronous (to avoid tearing), and there is no other way (or only less performant ways) we could implement this without `useSyncExternalStore`.

**Comment by OliverJAsh** [comment]
That's what I thought. Do you know if there's been any further discussion in the React team about this? I understand the constraints but this seems like a pretty fundamental flaw to me.

**Comment by markerikson** [comment]
Purely off the top of my head, this is basically the React team's intent.  The only way React can fully _know_ that a state update is interruptible / low-pri is if it's built-in React state, in which case React controls when the render pass happens and when that queued state update gets applied.  React has no control over any external store, and those update synchronously, so there's no way to tie together a future pending React update with the external store somehow applying the change to itself later when React is ready to render.  So, React is forced to treat all external state updates and renders as synchronous and blocking too.

**Comment by JAAAAAEMKIM** [comment]
I'm suffering from the same issue too. This sync work blocks rendering for so much ms. Are there at least any recommended workaround? 👀
What I'm trying is using `useDeferredValue` to defer from original state whenever I need transitions.

**Comment by OliverJAsh** [comment]
In our case we were able to workaround the problem by moving the state out of Redux and into context. This does make me question the long term usage of Redux state in React. 🤔

**Comment by palnes** [comment]
There is ongoing work from React on this, as per their latest blog post: 
https://react.dev/blog/2025/04/23/react-labs-view-transitions-activity-and-more#concurrent-stores

**Comment by markerikson** [comment]
Yeah, I don't have any further details on what these "concurrent stores" will entail, but at first glance it _sounds_ like it's meant to be a sort of "`uSES` but concurrent friendly" API hopefully?  Assuming that's the case, we'll definitely try to make use of that once it's available.

**Comment by markerikson** [comment]
@einarq : yeah, I'm keeping a close eye on any mentions of concurrent stores. Hoping to talk with the React team at ReactConf and afterwards to get a better understanding of what the API is supposed to do and how we might be able to use it.

**Comment by veksa** [comment]
Coming at this from a trading-app angle, with an approach that's different from "wait for concurrent stores" — because I think part of this is fixable in userland today.

There are two costs bundled into "Redux updates render synchronously": the **render** (sync lane, needs concurrent stores — agreed, can't touch that in userland), and the **notification fan-out** — walking every subscriber (`checkForUpdates` → selector → equality) before any render. That's `O(subscribers)` in one uninterrupted task, and *that* we can slice.

The idea: walk the subscriber list in ~5ms chunks and yield between them, using React's own `scheduler` primitive (`shouldYield()` on a 5ms budget, `startTime` per chunk, one reused `MessageChannel`). The store stays fully synchronous — `getSnapshot` and state are untouched, `dispatch` still commits immediately. Only *when each subscriber is told to re-check* becomes async, which is safe with uSES: it re-reads the snapshot on render and re-renders if it changed, so a subscriber told a few ms late just renders a few ms late against the same committed snapshot — no tearing.

I added a realistic scenario to react-redux-benchmarks for exactly this shape — a forex board, 10k rows, a handful ticking per frame, each with a non-memoized selector (the existing `many-components-*` use trivial selectors, `derived-selectors` is memoized): reduxjs/react-redux-benchmarks#39. Published `9.2.0` vs the sliced build, headless Chromium:

| metric | `9.2.0` | sliced |
|---|--:|--:|
| **synchronous `dispatch` (avg)** — the notification fan-out | `11.6 ms` | **`1.8 ms`** |

The ~10ms fan-out leaves the sync path; the rest is just the reducer. It's not just my scenario — the existing `many-components-*` scenarios move the same way:

| scenario | metric | `9.2.0` | sliced |
|---|---|--:|--:|
| `many-components-same-slice` (30k) | sync `dispatch` / p95 commit | `6.2 ms` / `96 ms` | `0.1 ms` / `44 ms` |
| `many-components-many-slices` (5k) | sync `dispatch` | `3.3 ms` | `2.0 ms` |

And a standalone micro-bench isolating the pass shows longest task `265 ms → 5 ms` and worst-case input latency `266 ms → 5 ms` at 100k subscribers.

Honest scope: this helps *many cheap subscribers* (the common large-app case), not *few expensive renders* — React still won't slice one component's render. It's a stopgap for stable React 18/19, not a replacement for concurrent stores. (It also yields early on `navigator.scheduling.isInputPending()`.)

It's all working on a branch already — **opt-in**, so the default stays synchronous — re-entrancy-guarded so subscription order and exactly-once-per-pass still hold, with tests. I'd be glad to turn it into a PR, either as a `<Provider>` opt-in or a store enhancer alongside `autoBatchEnhancer`. Is that a direction you'd consider?

**Comment by markerikson** [comment]
Subscriber notification cost is exactly the scenario I'm already trying to address over in https://github.com/reduxjs/react-redux/pull/2318. I can take a look at your benchmarks scenario in a bit and see how it performs with those changes, but I'd also appreciate it if you could try it the preview build in your own app and see how it works for both perf and correctness.

**Comment by veksa** [comment]
Thanks @markerikson! Rather than guess how these compare, I built both into the harness and ran them side by side — stock `9.2.0`, my time-sliced `optimize-subscription`, and your #2318 `signals` build, same scenarios and machine. Two scenarios that hit the fan-out from opposite ends. All medians of 5 runs, 20s each, headless Chromium, React 19 prod profiling build; ranges are min–max across those runs.

**`price-ticker`** — 10k rows, ~30 tick/frame, non-memoized per-row selector. Sparse: almost nothing changes per tick, but every subscriber re-runs.

| build | sync `dispatch` (median) | range | dispatches / 20s | mount | total script |
|---|--:|--:|--:|--:|--:|
| `9.2.0` | 5.92 ms | 5.65–9.70 | 1323 | 27.7 ms | 14794 ms |
| **`optimize-subscription`** | **1.28 ms** | 1.25–2.29 | 931 | 33.5 ms | 14587 ms |
| `signals` (#2318) | 3.26 ms | 3.22–3.43 | 1430 | **926 ms** | **13761 ms** |

**`many-components-same-slice`** — 5k components on a trivial `state.counter`. Dense: everyone depends on the one path that changes.

| build | sync `dispatch` (median) | range | mount | avg commit |
|---|--:|--:|--:|--:|
| `9.2.0` | 0.64 ms | 0.59–0.67 | 17.5 ms | 4.90 ms |
| **`optimize-subscription`** | **0.08 ms** | 0.07–0.08 | 19.6 ms | 5.88 ms |
| `signals` (#2318) | 1.32 ms | 1.18–1.88 | 16.9 ms | **3.15 ms** |

And where the sync time actually goes — an instrumented build timing the notify pass at the Redux-listener boundary (same hook for all three, so it's comparable). Cumulative synchronous notify over 20s:

| build | `price-ticker` | `many-components` |
|---|--:|--:|
| `9.2.0` | 8704 ms | 259 ms |
| **`optimize-subscription`** | **4 ms** | **2 ms** |
| `signals` (#2318) | 2823 ms | 230 ms |

Both help, but the shape of the win is very different, and mount is where it starts.

Signals pays ~926 ms at mount vs ~28 ms — building the proxy/registry for 10k tracked selectors before the app renders a frame. That's ~33× stock, and it lands squarely on the "big update on app load" case @mikeduminy raised earlier in this thread, which is a big part of what this issue is about. On dense updates it also regresses (1.32 vs 0.64 ms): when every subscriber depends on the path that changed there's nothing to prune, so the proxy/diff overhead is pure cost.

Its one real advantage is narrow and *amortized*, not free. On sparse updates it touches only changed paths, so each steady-state dispatch is cheaper — enough that over this 20s continuous-tick window it even nets lower total script (13.8s vs 14.8s) and higher throughput (1430). But you only come out ahead after enough sparse ticks to pay back that ~900 ms mount — order of ~400 dispatches / several seconds of nonstop ticking here. For app load, short interactions, or dense updates it's a net loss. (I'd stress it's a draft, so the mount cost may be tuneable — but it's structural to building per-selector tracking up front.)

Slicing has no such break-even. It's a drop-in, opt-in, mount-neutral (~34 vs ~28 ms), and it drops sync dispatch in both regimes from the very first dispatch (5.92→1.28, 0.64→0.08) because it doesn't care about update shape — it just refuses to run the whole fan-out in one uninterruptible task, so ~4 ms stays synchronous and the rest is deferred.

Honest caveat on mine: it cuts *latency*, not *total work*. The deferred pass still costs main-thread time, so under a dispatch firehose throughput dips (931 vs 1323 stock in the window) and per-commit render is a hair slower (5.88 vs 4.90 ms). But there's no upfront tax and nothing to amortize — it helps from frame one.

So they're really on different axes: signals attacks *how much* fan-out work exists (and only pays off on sustained sparse updates, once you've eaten the mount), slicing attacks *when* it runs (a smaller but unconditional win). For the load-time blocking this issue actually reports, slicing is deployable on stable React today with no tax; signals is the more invasive change whose mount would need to come down before it fits that case. They could even compose — a slicing enhancer under a signals provider.

Two things so it's not taken on faith: the signals build is #2318 head, and its `exports` already alias `Provider`/`useSelector` to the signal versions, so the *identical* scenario code runs the real signals path — no per-build edits. And both selectors here return a primitive, so signals' object-identity caveat never fires; if anything that makes this a best case for it, not one rigged against it. (Scenario: reduxjs/react-redux-benchmarks#39.) And a real large app is if anything a harder case than this synthetic one — more subscribers and heavier non-memoized selectors — so the fan-out these numbers isolate should show up at least as strongly there, not less.

**Comment by markerikson** [comment]
@veksa fyi I did some major work last night after I posted that comment to optimize mounting behavior and cases of primitive arrays, and that branch is now a win in that scenario.  Mounting is _modestly_ slower than stock 9.2, but more like 1.5x.  Also been improving a lot of edge case handling as well.  Will push that later today.

The "cuts latency but not total work" is my first concern about slicing subscribers.  I don't see that as sufficient benefit.  Not ruling it out right away, but the real problem is the total amount of work being done, which is what I'm trying to tackle.

**Comment by markerikson** [comment]
@veksa pushed the updates that optimize primitive arrays and other cases, and `useSignalSelector` is now a win on the ticket benchmark - see https://github.com/reduxjs/react-redux/pull/2318#issuecomment-5127081211 .
